crates/binding_wasm/src/lib.rs

```diff
@@ -1,14 +1,146 @@
-pub fn add(left: usize, right: usize) -> usize {
-    left + right
+use std::{cell::RefCell, io::BufWriter, path::PathBuf, str::FromStr, sync::Arc};
+
+use anyhow::{anyhow, Result};
+use serde::{Deserialize, Serialize};
+use stc_ts_env::{BuiltIn, Env, ModuleConfig, Rule, StableEnv};
+use stc_ts_type_checker::{
+    loader::{LoadFile, ModuleLoader},
+    Checker,
+};
+use swc_common::{
+    errors::{EmitterWriter, Handler},
+    FileName, Globals, SourceFile, SourceMap, GLOBALS,
+};
+use swc_ecma_ast::EsVersion;
+use swc_ecma_loader::resolve::Resolve;
+use swc_ecma_parser::Syntax;
+
+const ENTRY_FILENAME: &str = "binding-wasm-entry-module";
+
+#[cfg(target_arch = "wasm32")]
+type Out = wasm_bindgen::JsValue;
+#[cfg(not(target_arch = "wasm32"))]
+type Out = CheckAndAnnotateOutput;
+
+#[cfg_attr(target_arch = "wasm32", wasm_bindgen::prelude::wasm_bindgen)]
+pub fn check_and_annotate_types(src: &str) -> Out {
+    let globals = Arc::<Globals>::default();
+    let cm = Arc::new(SourceMap::default());
+    let buf = Arc::new(RefCell::new(BufWriter::new(Vec::new())));
+    let env = GLOBALS.set(&globals, || {
+        Env::new(
+            StableEnv::new(),
+            Rule { ..Default::default() },
+            EsVersion::latest(),
+            ModuleConfig::None,
+            Arc::new(BuiltIn::default()),
+        )
+    });
+    let emitter = Box::new(EmitterWriter::new(
+        Box::new(PassThroughWriter { buf: buf.clone() }),
+        Some(cm.clone()),
+        false,
+        false,
+    ));
+    let handler = { Arc::new(Handler::with_emitter(true, false, emitter)) };
+    let (types, errors) = GLOBALS.set(&globals, || {
+        let loader = SingleEntryLoader::new(cm.as_ref(), src);
+        let mut checker = Checker::new(
+            cm.clone(),
+            handler.clone(),
+            env.clone(),
+            None,
+            Box::new(ModuleLoader::new(cm, env, VoidResolver, loader)),
+        );
+
+        let entry_module = checker.check(Arc::new(FileName::Real(PathBuf::from_str(ENTRY_FILENAME).unwrap())));
+        let types = checker.get_types(entry_module);
+
+        (types, checker.take_errors())
+    });
+    let types = types.map(|types| format!("{:?}", types)).unwrap_or_else(|| "".to_owned());
+
+    for err in &errors {
+        err.emit(&handler);
+    }
+
+    let out = CheckAndAnnotateOutput {
+        error: String::from_utf8_lossy(buf.borrow().get_ref()).into_owned(),
+    };
+
+    #[cfg(target_arch = "wasm32")]
+    {
+        serde_wasm_bindgen::to_value(&out).unwrap()
+    }
+
+    #[cfg(not(target_arch = "wasm32"))]
+    {
+        out
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct CheckAndAnnotateOutput {
+    error: String,
+}
+
+pub struct PassThroughWriter {
+    pub buf: Arc<RefCell<BufWriter<Vec<u8>>>>,
```

Comment:
This should be `Arc<Mutex<Vec<u8>>>`.

 - `BufWriter<Vec<u8>>` has nothing good over `Vec<u8>` becuase it's already in memory.
 - `Arc<RefCell<T>>` is unsound, and it's `unsafe.`